File: webworker/src/lib/Resolution.js
import Cartographic from "./Cartographic"
import EllipsoidGeodesic from "./EllipsoidGeodesic"
import DeveloperError from "./DeveloperError.js";
/**
 * 获取某一层在赤道处的地图分辨率
 * @param {number} level 
 */
function getLevelResolution(level, tileSize) {
    const xNumAtZeroLevel = 2
    const yNumAtZeroLevel = 1
    //6378137.0, 6378137.0, 6356752.3142451793
    const earthRadius = 6378137
    return (2 * Math.PI * earthRadius) / (tileSize * (xNumAtZeroLevel << level))
}
const MAX_LEVEL = 30
/**
 * 计算分辨率
 * quality的意思是矩形的宽度等于几个256，用来确定多边形的分辨率。分辨率=矩形宽度(米)/(quality*256像素)
 */
export default class Resolution {
    constructor(tileSize = 256, quality = 0.8) {
        this.tileSize = tileSize
        this.quality = quality
        this.geodesic = new EllipsoidGeodesic();
        const xNumAtZeroLevel = 2
        const earthRadius = 6378137
        this.metersPerPixelAtZeroLevel = (2 * Math.PI * earthRadius) / (this.tileSize * xNumAtZeroLevel)
    }
    get tileSizeWithQuality() {
        return Math.ceil(this.tileSize * this.quality)
    }
    computeLevel(tileRectangle, epsg, maxLevel) {
        const midLatitude = (tileRectangle.south + tileRectangle.north) / 2
        const startPoint = new Cartographic(tileRectangle.west, midLatitude)
        const endPoint = new Cartographic(tileRectangle.east, midLatitude)
        this.geodesic.setEndPoints(startPoint, endPoint);
        let level = (this.metersPerPixelAtZeroLevel * Math.cos(midLatitude)) / (this.geodesic.surfaceDistance / this.tileSizeWithQuality)
        level = Math.log(level) / Math.log(2)
        level = Math.round(level) | 0
        if (level > maxLevel) {
            level == maxLevel
        }
        return level
    }
    computeLevel2(tileRectangle, epsg, maxLevel) {
        let tempScheme = null;
        let initRadius = Math.PI;

        if (epsg === 3857) {
            initRadius = Math.PI * 2;
        }
        if (epsg === 4326) {
            initRadius = Math.PI;
        }

        let boundingBox = tileRectangle
        let west_north_point = [boundingBox.west, boundingBox.north];
        let east_south_point = [boundingBox.east, boundingBox.south];

        let gap_lon = boundingBox.east - boundingBox.west;
        let level = 0;

        let isSkip = false;
        while (true && !isSkip) {
            let currentLevelGap = initRadius / Math.pow(2, level);
            let nextLevelGap = initRadius / Math.pow(2, level + 1);

            if (gap_lon <= currentLevelGap && gap_lon >= nextLevelGap) {
                level = level + 1;
                isSkip = true;
            } else {
                level++;
            }
        }
        if (level > maxLevel) {
            level = maxLevel;
        }
        return level
    }
}

File: rust/src/resolution.rs
use std::f64::consts::PI;

use crate::{
    cartographic::Cartographic, ellipsoidal_geodesic::EllipsoidGeodesic, rectangle::Rectangle,
};

pub struct Resolution {
    tile_size: u32,
    quality: f64,
    geodesic: EllipsoidGeodesic,
    metersPerPixelAtZeroLevel: f64,
}
impl Resolution {
    pub fn new(tile_size: Option<u32>, quality: Option<f64>) -> Self {
        let tileSize = tile_size.unwrap_or(256);
        let quality = quality.unwrap_or(0.8);
        let xNumAtZeroLevel = 2.0;
        let earthRadius = 6378137.0;
        let metersPerPixelAtZeroLevel =
            (2.0 * PI * earthRadius) / ((tileSize as f64) * xNumAtZeroLevel);
        return Self {
            tile_size: tileSize,
            quality: quality,
            geodesic: EllipsoidGeodesic::default(),
            metersPerPixelAtZeroLevel: metersPerPixelAtZeroLevel,
        };
    }
    pub fn tileSizeWithQuality(self: &Self) -> f64 {
        return (self.tile_size as f64) * self.quality;
    }
    pub fn computeLevel(
        self: &mut Self,
        tile_rectangle: &Rectangle,
        epsg: u32,
        max_level: u32,
    ) -> u32 {
        let midLatitude = (tile_rectangle.south + tile_rectangle.north) / 2.0;
        let startPoint = Cartographic::new(tile_rectangle.west, midLatitude, 0.0);
        let endPoint = Cartographic::new(tile_rectangle.east, midLatitude, 0.0);
        self.geodesic.setEndPoints(startPoint, endPoint);
        let tileSizeWithQuality = self.tileSizeWithQuality();
        let mut level = (self.metersPerPixelAtZeroLevel * midLatitude.cos())
            / (self.geodesic.get_surface_distance() / tileSizeWithQuality);
        level = level.ln() / 2f64.ln();
        let mut rounded = level.round() as u32;
        if rounded > max_level {
            rounded = max_level;
        }
        return rounded | 0;
    }
    pub fn computeLevel2(
        self: &Self,
        tile_rectangle: &Rectangle,
        epsg: u32,
        max_level: u32,
    ) -> u32 {
        let east = tile_rectangle.east;
        let west = tile_rectangle.west;
        let init_radius = std::f64::consts::PI;
        let mut level: u32 = 0;
        let gap_lon = east - west;
        loop {
            let current_level_gap = init_radius / ((2u32.pow(level)) as f64);
            let next_level_gap = init_radius / ((2u32.pow(level + 1)) as f64);
            if gap_lon <= current_level_gap && gap_lon >= next_level_gap {
                level = level + 1;
                break;
            } else {
                level = level + 1;
            }
        }
        if level > max_level {
            level = max_level;
        }
        return level;
    }
}
